pkg/runner/test_runner.go

```diff
@@ -6,282 +6,796 @@
+				break loop
+			}
+			if err = tr.reportTargetResult(ctx, ss, res.Target, res.Err); err != nil {
+				break loop
+			}
+		case <-ctx.Done():
+			ss.log.Debugf("%s: canceled 3, draining", ss)
+			for {
+				select {
+				case <-outCh:
+					break loop
+				case <-ss.errCh:
+					break loop
+				case <-time.After(tr.shutdownTimeout):
+					tr.mu.Lock()
+					if ss.runErr == nil {
+						ss.runErr = &cerrors.ErrTestStepsNeverReturned{}
+					}
+					tr.mu.Unlock()
+				}
+			}
+		}
+	}
+	tr.mu.Lock()
+	defer tr.mu.Unlock()
+	if ss.runErr == nil && err != nil {
+		ss.runErr = err
+	}
+	if ss.stepRunning && ss.runErr == nil {
+		// This means that plugin closed its channels before leaving.
+		ss.runErr = &cerrors.ErrTestStepClosedChannels{StepName: ss.sb.TestStepLabel}
+	}
+	ss.readerRunning = false
+	ss.log.Debugf("%s: step reader finished, %t %t %v", ss, ss.stepRunning, ss.readerRunning, ss.runErr)
+	tr.cond.Signal()
 }
 
-// CompletedTargets returns a map that associates each target with its returning error.
-// If the target succeeded, the error will be nil
-func (r *State) CompletedTargets() map[*target.Target]error {
-	return r.completedTargets
+// checkStepRunnersLocked checks if any step runner has encountered an error.
+func (tr *TestRunner) checkStepRunners() error {
+	for _, ss := range tr.steps {
+		if ss.runErr != nil {
+			return ss.runErr
+		}
+	}
+	return nil
 }
 
-// CompletedRouting returns a map that associates each routing block with its returning error.
-// If the routing block succeeded, the error will be nil
-func (r *State) CompletedRouting() map[string]error {
-	return r.completedRouting
+// runMonitor monitors progress of targets through the pipeline
+// and closes input channels of the steps to indicate that no more are expected.
+// It also monitors steps for critical errors and cancels the whole run.
+// Note: input channels remain open when cancellation is requested,
+// plugins are expected to handle it explicitly.
+func (tr *TestRunner) runMonitor() error {
+	tr.log.Debugf("monitor: active")
+	tr.mu.Lock()
+	defer tr.mu.Unlock()
+	// First, compute the starting step of the pipeline (it may be non-zero
+	// if the pipeline was resumed).
+	minStep := len(tr.steps)
+	for _, ts := range tr.targets {
+		if ts.CurStep < minStep {
+			minStep = ts.CurStep
+		}
+	}
+	if minStep < len(tr.steps) {
+		tr.log.Debugf("monitor: starting at step %s", tr.steps[minStep])
+	}
+
+	// Run the main loop.
+	pass := 1
+	var runErr error
+loop:
+	for step := minStep; step < len(tr.steps); pass++ {
+		ss := tr.steps[step]
+		tr.log.Debugf("monitor pass %d: current step %s", pass, ss)
+		// Check if all the targets have either made it past the injection phase or terminated.
+		ok := true
+		for _, ts := range tr.targets {
+			tr.log.Debugf("monitor pass %d: %s: %s", pass, ss, ts)
+			if ts.resCh == nil { // Not running anymore
+				continue
+			}
+			if ok && (ts.CurStep < step || ts.CurPhase < targetStepPhaseRun) {
```

Comment:
This truly deserves a comment. I don't understand this check, why we break down below, and why `if !ok`, we just wait for condition variable to be signaled and continue looping into the following step.

Reply:
the comment above explains the purpose of the entire loop - check if all the targets have either made it past the injection phase or terminated.
the first condition checks if the target has terminated or not (if terminated, there will not be a result channel).
this one checks if it made past the target injection phase.
it's perfectly fine for this to not be the case, in this case we just wait for things to change (for condition to be signaled).

i simplified the check a bit, since we break out on first !ok, there's no reason to test for it (a leftover from previous implementation where we'd iterate all targets anyway).